app/helpers/tabs_helper.rb

```diff
@@ -1,107 +1,88 @@
 # encoding: utf-8
+# html used in tabsets
 module TabsHelper
-
   # Short-hand to render shared tab_set partial for a given set of links.
   def draw_tab_set(links)
-    render(partial: "/shared/tab_set", locals: {links: links})
+    render(partial: "/shared/tab_set", locals: { links: links })
   end
 
-  # Draw the cutesy eye icons in the upper right side of screen.  It does it
-  # by creating a "right" tab set.  Thus this must be called in the header of
-  # the view and must not actually be rendered.  Typical usage would be:
-  #
-  #   # At top of view:
-  #   <%
-  #     # Specify the page's title.
-  #     @title = "Page Title"
-  #
-  #     # Define set of linked text tabs for top-left.
-  #     new_tab_set do
-  #       add_tab("Tab Label One", link: args, ...)
-  #       add_tab("Tab Label Two", link: args, ...)
-  #       ...
-  #     end
-  #
-  #     # Draw interest icons in the top-right.
-  #     draw_interest_icons(@object)
-  #   %>
-  #
-  # This will cause the set of three icons to be rendered floating in the
-  # top-right corner of the content portion of the page.
-  #
-  def draw_interest_icons(object)
-    if @user
-      type = object.type_tag
+  # assemble HTML for "tabset" for show_observation
+  # actually a list of links and the interest icons
+  def show_observation_tabset(obs, user)
+    tabs = []
 
-      # Create link to change interest state.
-      def interest_link(label, object, state) # :nodoc:
-        link_with_query(label,
-          controller: :interest,
-          action: :set_interest,
-          id: object.id,
-          type: object.class.name,
-          state: state
-        )
-      end
+    tabs << google_images_and_dist_map_links(obs)
+    tabs << general_questions_link(obs, user)
+    tabs << notifications_link(obs, user)
+    tabs << manage_lists_link(obs, user)
+    tabs << map_link
+    tabs << obs_change_links(obs)
 
-      # Create large icon image.
-      def interest_icon_big(type, alt) # :nodoc:
-        image_tag("#{type}2.png",
-          alt: alt,
-          width: "50px",
-          height: "50px",
-          class: "interest_big",
-          title: alt
-        )
-      end
+    tabs << draw_interest_icons(obs)
 
-      # Create small icon image.
-      def interest_icon_small(type, alt) # :nodoc:
-        image_tag("#{type}3.png",
-          alt: alt,
-          width: "23px",
-          height: "23px",
-          class: "interest_small",
-          title: alt
-        )
-      end
+    { pager_for: obs, right: draw_tab_set(tabs.reject(&:empty?)) }
+  end
 
-      def interest_tab(img1, img2, img3)
-        content_tag(:div, img1 + safe_br + img2 + img3, class: "interest-eyes")
-        # content_tag(:div, img1 + safe_br + img2 + img3, style: "position: absolute; top: 0;")
-      end
+  def google_images_and_dist_map_links(obs)
+    return unless obs.try(:name).known?
```

Comment:
Yeah, all three are similar... but I think (1) is the most clear and robust.

1) return unless obs && obs.name && obs.name.known?
2) return unless obs.name.known? rescue false
3) return unless obs.try(:name).known?

What happens in (3) if obs.name is nil for example?  And how is (2)
parsed?  Does it return if it throws an error, or does the statement
evaluate to false and continue to the next line?

On Thu, Oct 15, 2015 at 5:40 PM, Joseph D. Cohen <notifications@github.com>
wrote:

> In app/helpers/tabs_helper.rb
> <https://github.com/MushroomObserver/mushroom-observer/pull/187#discussion_r42198787>
> :
>
> >
> > -      def interest_tab(img1, img2, img3)
> > -        content_tag(:div, img1 + safe_br + img2 + img3, class: "interest-eyes")
> > -        # content_tag(:div, img1 + safe_br + img2 + img3, style: "position: absolute; top: 0;")
> > -      end
> > +  def google_images_and_dist_map_links(obs)
> > +    return unless obs.try(:name).known?
>
> What about the predecessor: if name && !name.unknown?
> … <#1506e16d74d04cfb_>
> *--Joe Cohen*
> On Thu, Oct 15, 2015 at 5:35 PM, Jason Hollinger <notifications@github.com>
> wrote: In app/helpers/tabs_helper.rb <#187 (comment)
> <https://github.com/MushroomObserver/mushroom-observer/pull/187#discussion_r42198538>>
> : > > - def interest_tab(img1, img2, img3) > - content_tag(:div, img1 +
> safe_br + img2 + img3, class: "interest-eyes") > - # content_tag(:div, img1
> + safe_br + img2 + img3, style: "position: absolute; top: 0;") > - end > +
> def google_images_and_dist_map_links(obs) > + return unless
> obs.try(:name).known? Not I. I never saw the "obj.try(:method)" trick. I
> use "obj.method rescue nil" instead (if I'm being lazy). …
> <#1506e1285d7e8e4a_> On Thu, Oct 15, 2015 at 5:29 PM, Joseph D. Cohen <
> notifications@github.com> wrote: In app/helpers/tabs_helper.rb <#187
> <https://github.com/MushroomObserver/mushroom-observer/pull/187>
> (comment) <#187 (comment)
> <https://github.com/MushroomObserver/mushroom-observer/pull/187#discussion_r42198213>>>
> : > > - def interest_tab(img1, img2, img3) > - content_tag(:div, img1 +
> safe_br + img2 + img3, class: "interest-eyes") > - # content_tag(:div, img1
> + safe_br + img2 + img3, style: "position: absolute; top: 0;") > - end > +
> def google_images_and_dist_map_links(obs) > + return unless
> obs.try(:name).known? 1. return unless obs.try(:name).known? is my
> rewriting of master branch show_observation.html.erb, line 13, if name &&
> !name.unknown?. I was also puzzled about the first condition, and thought
> it might be best to throw an error. But I eventually decided not to monkey
> with it. Perhaps @pellaea <https://github.com/pellaea> <
> https://github.com/pellaea> < https://github.com/pellaea> knows? Maybe at
> one time there was a Name fixture without a name? But now all tests pass if
> the first condition is removed, so I will remove it. — Reply to this email
> directly or view it on GitHub <
> https://github.com/MushroomObserver/mushroom-observer/pull/187/files#r42198213>
> . — Reply to this email directly or view it on GitHub <
> https://github.com/MushroomObserver/mushroom-observer/pull/187/files#r42198538>
> .
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/MushroomObserver/mushroom-observer/pull/187/files#r42198787>
> .
>